DOCS-#3678: Improve documentation for `pandas on Ray` execution (#3679)
modin-project/modin: docs/flow/modin/core/execution/ray/implementations/pandas_on_ray/index.rst

PandasOnRay Execution
=====================

> Flows which a user’s queries for data transformation, data ingress and egress pass through the Modin components

Author: We are trying to get rid of `backend` word. Since the recent restructure we use `execution` for describing the whole name of a component that consists of a storage format and an engine. That might also be `execution system` or `execution backend`.

for pandas on Ray execution are shown below.

To enable the execution, please refer to :doc:`pandas on Ray </UsingPandasonRay/index>` usage section.

Data Transformation
'''''''''''''''''''

.. image:: /img/pandas_on_ray_data_transform.svg
   :align: center

When a user calls any of :py:class:`~modin.pandas.dataframe.DataFrame` API, a query starts forming at the `API` layer
to be executed at the `Execution` layer. The `API` layer’s responsibility is to process the query in an appropriate way
for the final result (like, detect whether ``DataFrame`` or ``Series`` should be returned) and to ensure clean input to
:py:class:`~modin.core.storage_formats.pandas.query_compiler.PandasQueryCompiler` (like, validate a parameter of the call
to define specific values for it to pass that information to the query compiler).

> :py:class:`~modin.core.storage_formats.pandas.query_compiler.PandasQueryCompiler` is in charge of

Reviewer: ```suggestion
The :py:class:`~modin.core.storage_formats.pandas.query_compiler.PandasQueryCompiler` is responsible for
```

> processing passed query in a way that is conforming for the subset of data (cell-wise or axis-wise partition) of `pandas`

Reviewer: ```suggestion
processing the query, received from the :py:class:`~modin.pandas.dataframe.DataFrame` API layer,  to determine how to apply it to a subset of the data - either cell-wise or along an axis-wise partition backed by the `pandas`
```

storage format and mapping the query to one of the :doc:`Core Operators </flow/modin/core/dataframe/algebra>` of
:py:class:`~modin.core.execution.ray.implementations.pandas_on_ray.dataframe.dataframe.PandasOnRayDataframe` that inherits
generic functionality from ``GenericRayDataframe`` and :py:class:`~modin.core.dataframe.pandas.dataframe.dataframe.PandasDataframe`.

PandasOnRay Dataframe implementation
------------------------------------

Modin implements ``Dataframe``, ``PartitionManager``, ``AxisPartition`` and ``Partition`` classes

> specific for ``PandasOnRay`` execution:

Reviewer: ```suggestion
specifically for the ``PandasOnRay`` backend:
```

PandasOnRay Execution
=====================

Queries that perform data transformation, data ingress or data egress using the `pandas on Ray` execution
pass through the Modin components detailed below.

To enable the `pandas on Ray` execution, please refer to the usage section in :doc:`pandas on Ray </UsingPandasonRay/index>`.

Data Transformation
'''''''''''''''''''

.. image:: /img/pandas_on_ray_data_transform.svg
   :align: center

When a user calls any :py:class:`~modin.pandas.dataframe.DataFrame` API, a query starts forming at the `API` layer
to be executed at the `Execution` layer. The `API` layer is responsible for processing the query appropriately,

> for example, determining whether the final result should be a ``DataFrame`` or ``Series`` object, and sanitizing the input to the

Reviewer 2: Shouldn't there be links here to `DataFrame` and `Series`?

Author: That doesn't make a lot of sense. We refer to DataFrame and Series as general objects here.

:py:class:`~modin.core.storage_formats.pandas.query_compiler.PandasQueryCompiler`, e.g. validating a parameter from the query
and defining specific intermediate values to provide more context to the query compiler.
The :py:class:`~modin.core.storage_formats.pandas.query_compiler.PandasQueryCompiler` is responsible for
processing the query, received from the :py:class:`~modin.pandas.dataframe.DataFrame` `API` layer,
to determine how to apply it to a subset of the data - either cell-wise or along an axis-wise partition backed by the `pandas`
storage format and mapping the query to one of the :doc:`Core Operators </flow/modin/core/dataframe/algebra>` of
:py:class:`~modin.core.execution.ray.implementations.pandas_on_ray.dataframe.dataframe.PandasOnRayDataframe` that inherits

> generic functionality from ``GenericRayDataframe`` and :py:class:`~modin.core.dataframe.pandas.dataframe.dataframe.PandasDataframe`.

Author: We do not have `GenericRayDataframe` class.

> generic functionality from ``GenericRayDataframe`` and :py:class:`~modin.core.dataframe.pandas.dataframe.dataframe.PandasDataframe`.

Reviewer 2: But we have doc, which I mentioned above with description of `GenericRayDataframePartitionManager`. It can help with fast finding info about `GenericRayDataframe` is mean in that context (becaise it isn't a class).

> generic functionality from ``GenericRayDataframe`` and :py:class:`~modin.core.dataframe.pandas.dataframe.dataframe.PandasDataframe`.

Author: `GenericRayDataframe` is a class here and I would prefer to have a direct link to it once we add an implementation of the class.

> generic functionality from ``GenericRayDataframe`` and :py:class:`~modin.core.dataframe.pandas.dataframe.dataframe.PandasDataframe`.

Reviewer 2: Ok. In this case we should don't forget to do this (mention of an issue related to adding of `GenericRayDataframe` in the comment in this thread will be enough)

> generic functionality from ``GenericRayDataframe`` and :py:class:`~modin.core.dataframe.pandas.dataframe.dataframe.PandasDataframe`.

Author: Added todo.

PandasOnRay Execution
=====================

Queries that perform data transformation, data ingress or data egress using the `pandas on Ray` execution
pass through the Modin components detailed below.

> To enable the `pandas on Ray` execution, please refer to the usage section in :doc:`pandas on Ray </UsingPandasonRay/index>`.

Reviewer: ```suggestion
To enable `pandas on Ray` execution, please refer to the usage section in :doc:`pandas on Ray </UsingPandasonRay/index>`.
```

Data Transformation
'''''''''''''''''''

.. image:: /img/pandas_on_ray_data_transform.svg
   :align: center

When a user calls any :py:class:`~modin.pandas.dataframe.DataFrame` API, a query starts forming at the `API` layer
to be executed at the `Execution` layer. The `API` layer is responsible for processing the query appropriately,

> for example, determining whether the final result should be a ``DataFrame`` or ``Series`` object, and sanitizing the input to the

Reviewer: ```suggestion
for example, determining whether the final result should be a ``DataFrame`` or ``Series`` object. This layer is also responsible for sanitizing the input to the
```

:py:class:`~modin.core.storage_formats.pandas.query_compiler.PandasQueryCompiler`, e.g. validating a parameter from the query
and defining specific intermediate values to provide more context to the query compiler.
The :py:class:`~modin.core.storage_formats.pandas.query_compiler.PandasQueryCompiler` is responsible for
processing the query, received from the :py:class:`~modin.pandas.dataframe.DataFrame` `API` layer,
to determine how to apply it to a subset of the data - either cell-wise or along an axis-wise partition backed by the `pandas`

> storage format and mapping the query to one of the :doc:`Core Operators </flow/modin/core/dataframe/algebra>` of

Reviewer: ```suggestion
storage format. The :py:class:`~modin.core.storage_formats.pandas.query_compiler.PandasQueryCompiler` maps the query to one of the :doc:`Core Algebra Operators </flow/modin/core/dataframe/algebra>` of
```

> :py:class:`~modin.core.execution.ray.implementations.pandas_on_ray.dataframe.dataframe.PandasOnRayDataframe` that inherits

Reviewer: ```suggestion
the :py:class:`~modin.core.execution.ray.implementations.pandas_on_ray.dataframe.dataframe.PandasOnRayDataframe` which inherits
```

> generic functionality from ``GenericRayDataframe`` and :py:class:`~modin.core.dataframe.pandas.dataframe.dataframe.PandasDataframe`.

Reviewer: ```suggestion
generic functionality from the ``GenericRayDataframe`` and the :py:class:`~modin.core.dataframe.pandas.dataframe.dataframe.PandasDataframe`.
```

    dataframe
    partitioning/partition
    partitioning/axis_partition
    partitioning/partition_manager

Data Ingress
''''''''''''

.. image:: /img/pandas_on_ray_data_ingress.svg
   :align: center

Data Egress
'''''''''''

.. image:: /img/pandas_on_ray_data_egress.svg
   :align: center

> When a user calls any of IO functions from ``modin.pandas.io`` module, `API` layer appeals to

Reviewer: ```suggestion
When a user calls any IO function from the ``modin.pandas.io`` module, the `API` layer queries the
```

> :py:class:`~modin.core.execution.dispatching.factories.dispatcher.FactoryDispatcher` that defines a factory specific for

Reviewer: ```suggestion
:py:class:`~modin.core.execution.dispatching.factories.dispatcher.FactoryDispatcher` which defines a factory specific for
```

> the execution, namely, :py:class:`~modin.core.execution.dispatching.factories.factories.PandasOnRayFactory`. The factory, in turn,

Reviewer: ```suggestion
the backend, namely, the :py:class:`~modin.core.execution.dispatching.factories.factories.PandasOnRayFactory`. The factory, in turn,
```

> holds :py:class:`~modin.core.execution.ray.implementations.pandas_on_ray.io.PandasOnRayIO` class whose responsibility is parallel read/write

Reviewer: ```suggestion
exposes the :py:class:`~modin.core.execution.ray.implementations.pandas_on_ray.io.PandasOnRayIO` class whose responsibility is to perform a parallel read/write
```

from/to a file.

> In the event of reading data from a CSV file, for example, :py:class:`~modin.core.execution.ray.implementations.pandas_on_ray.io.io.PandasOnRayIO` class forwards

Reviewer: ```suggestion
When reading data from a CSV file, for example, the :py:class:`~modin.core.execution.ray.implementations.pandas_on_ray.io.io.PandasOnRayIO` class forwards
```

the user query to the :meth:`~modin.core.io.text.CSVDispatcher._read` method of :py:class:`~modin.core.io.text.CSVDispatcher`, where the query's parameters are preprocessed

> to check if they are supported by the execution (otherwise, default pandas implementation is used to read the CSV file) and compute some metadata

Reviewer: ```suggestion
to check if they are supported by the execution (defaulting to pandas if they are not) and computes some metadata
```

common for all partitions to be read. Then, the file is split into row chunks and using this data remote tasks are launched on the Ray workers

> via :meth:`~modin.core.execution.ray.common.task_wrapper.RayTask.deploy` method of :py:class:`~modin.core.execution.ray.common.task_wrapper.RayTask` where :py:class:`~modin.core.storage_formats.pandas.parsers.PandasCSVParser` parses data

Reviewer: ```suggestion
via the :meth:`~modin.core.execution.ray.common.task_wrapper.RayTask.deploy` method of :py:class:`~modin.core.execution.ray.common.task_wrapper.RayTask`. On each Ray worker, the :py:class:`~modin.core.storage_formats.pandas.parsers.PandasCSVParser` parses data
```

> on each single Ray worker. After remote tasks are finished, additional result postprocessing is performed,

Reviewer: ```suggestion
After remote tasks are finished, additional result postprocessing is performed,
```

> and new query compiler with data read is returned.

Reviewer: ```suggestion
and a new query compiler with the data read is returned.
```

> In the event of writing data to a CSV file, for example, :py:class:`~modin.core.execution.ray.implementations.pandas_on_ray.io.PandasOnRayIO` processes

Reviewer: ```suggestion
When writing data to a CSV file, for example, the :py:class:`~modin.core.execution.ray.implementations.pandas_on_ray.io.PandasOnRayIO` processes
```

> the user query to execute it on Ray workers. Then, :py:class:`~modin.core.execution.ray.implementations.pandas_on_ray.io.PandasOnRayIO` appeals to

Reviewer: ```suggestion
the user query to execute it on Ray workers. Then, the :py:class:`~modin.core.execution.ray.implementations.pandas_on_ray.io.PandasOnRayIO` asks the
```

> :py:class:`~modin.core.execution.ray.implementations.pandas_on_ray.io.PandasOnRayDataframe` to be decomposed into row-wise partitions

Reviewer: ```suggestion
:py:class:`~modin.core.execution.ray.implementations.pandas_on_ray.io.PandasOnRayDataframe` to decompose the data into row-wise partitions
```

    partitioning/partition_manager

Data Ingress
''''''''''''

.. image:: /img/pandas_on_ray_data_ingress.svg
   :align: center

Data Egress
'''''''''''

.. image:: /img/pandas_on_ray_data_egress.svg
   :align: center

When a user calls any IO function from the ``modin.pandas.io`` module, the `API` layer queries the
:py:class:`~modin.core.execution.dispatching.factories.dispatcher.FactoryDispatcher` which defines a factory specific for
the execution, namely, the :py:class:`~modin.core.execution.dispatching.factories.factories.PandasOnRayFactory`. The factory, in turn,
exposes the :py:class:`~modin.core.execution.ray.implementations.pandas_on_ray.io.PandasOnRayIO` class
whose responsibility is to perform a parallel read/write from/to a file.

When reading data from a CSV file, for example, the :py:class:`~modin.core.execution.ray.implementations.pandas_on_ray.io.io.PandasOnRayIO` class forwards
the user query to the :meth:`~modin.core.io.text.CSVDispatcher._read` method of :py:class:`~modin.core.io.text.CSVDispatcher`, where the query's parameters are preprocessed
to check if they are supported by the execution (defaulting to pandas if they are not) and computes some metadata

> common for all partitions to be read. Then, the file is split into row chunks and using this data remote tasks are launched on the Ray workers

Reviewer: ```suggestion
common for all partitions to be read. Then, the file is split into row chunks, and this data is used to launch remote tasks on the Ray workers
```

via the :meth:`~modin.core.execution.ray.common.task_wrapper.RayTask.deploy` method of :py:class:`~modin.core.execution.ray.common.task_wrapper.RayTask`.
On each Ray worker, the :py:class:`~modin.core.storage_formats.pandas.parsers.PandasCSVParser` parses data.

> After remote tasks are finished, additional result postprocessing is performed,

Reviewer: ```suggestion
After the remote tasks are finished, additional result postprocessing is performed,
```
